qa/tasks/check_counter.py

```
@@ -124,7 +124,7 @@ def end(self):
                             expected_val = counter['expected_val']
                     else:
                         name = counter
-                        minval = 1
+                        minval = 0
```

Could you explain how this change fixes the issue (esp. for https://tracker.ceph.com/issues/70441)? The failed job(*) did not have the perf counter set to the desired value

```
2025-03-12T10:33:34.986 INFO:teuthology.orchestra.run.smithi162.stdout:        "req_rmsnap_latency": {
2025-03-12T10:33:34.986 INFO:teuthology.orchestra.run.smithi162.stdout:            "avgcount": 0,
2025-03-12T10:33:34.986 INFO:teuthology.orchestra.run.smithi162.stdout:            "sum": 0.000000000,
2025-03-12T10:33:34.986 INFO:teuthology.orchestra.run.smithi162.stdout:            "avgtime": 0.000000000
2025-03-12T10:33:34.986 INFO:teuthology.orchestra.run.smithi162.stdout:        },
```

The test expects the perf counter value to get incremented. I don't understand how changing the minval is going to fix the failure (it's going to hide the failure in the worst case).

(*): https://pulpito.ceph.com/vshankar-2025-03-12_09:22:14-fs-wip-vshankar-testing-20250306.043526-debug-testing-default-smithi/8183606/

---

@vshankar 
I'm not trying to hide the qa failure somehow.
Please read the descriptions in the commit and in the PR header and the changes made.
```
Most counters are having 0 default value in perf dump o/p.
This patch fixes the counters not 'seen' in the perf dump o/p.
```
It's not possible for check_counters.py to see the counter, if the `minval` doesn't match. Either `perf dump` output should be changed or `minval` should be changed. Since most counters starts with `0` default value, it's better change the code.

---

> @vshankar I'm not trying to hide the qa failure somehow. Please read the descriptions in the commit and in the PR header and the changes made.

I did and the changes didn't make sense to me.

> 
> ```
> Most counters are having 0 default value in perf dump o/p.
> This patch fixes the counters not 'seen' in the perf dump o/p.
> ```

Have you checked the failing test case? It expects `mds_server.req_rmsnap_latency.avgcount` to be at least 1 and the test involves removing snapshots. So, when snapshots are removed, the above counter value _should_ get incremented (even if one snapshot is removed). For some reason, in the failing test, it did not. So, how would reducing the default minval explain why the perf counter did't get incremented.

---

The case where `mds_server.req_rmsnap_latency.avgcount` would still be 0 is when no snapshots are removed. Did this happen in the test? If yes, then that explanation needs to be in the commit message of the fix that change minval to 0 or the other way would be to wait long enough in the test for the snapshot to be removed.

---

> Have you checked the failing test case? It expects mds_server.req_rmsnap_latency.avgcount to be at least 1 and the test involves removing snapshots. So, when snapshots are removed, the above counter value should get incremented (even if one snapshot is removed). For some reason, in the failing test, it did not. 

Yes, I've checked this. The initial value of `mds_server.req_rmsnap_latency.avgcount` is `0`, as you've mentioned^. That's why check_counters.py [failed](https://github.com/ceph/ceph/blob/main/qa/tasks/check_counter.py#L141) (`0 >= 1`) and the test didn't proceed to increment the counter.

> The case where mds_server.req_rmsnap_latency.avgcount would still be 0 is when no snapshots are removed. Did this happen in the test? If yes, then that explanation needs to be in the commit message of the fix that change minval to 0 or the other way would be to wait long enough in the test for the snapshot to be removed.

With respect to check_counters.py - yes. check_counters.py checks if the counters are greater than a particular value (minval). If it's not at the first occurrence, then the test would fail (`not seen`). It doesn't matter what happens next.
The `minval` should be fixed, as this issue most probably not hitting due to an earlier [bug](https://tracker.ceph.com/issues/67360). Now that the mismatch is revealed. 

I'll update the commit message with more details and I don't think we need to wait for snapshot to be removed, as check_counters just checks the presence of a counter.

---

> > Have you checked the failing test case? It expects mds_server.req_rmsnap_latency.avgcount to be at least 1 and the test involves removing snapshots. So, when snapshots are removed, the above counter value should get incremented (even if one snapshot is removed). For some reason, in the failing test, it did not.
> 
> Yes, I've checked this. The initial value of `mds_server.req_rmsnap_latency.avgcount` is `0`, as you've mentioned^. That's why check_counters.py [failed](https://github.com/ceph/ceph/blob/main/qa/tasks/check_counter.py#L141) (`0 >= 1`) and the test didn't proceed to increment the counter.

Yes, I know that. I wanted to understand the underlying reason as to why snap-schedule didn't remove snapshots. And the answer to that is in the ceph-mgr log:

```
2025-03-12T10:33:00.463+0000 7f3b739b2640  0 [snap_schedule DEBUG snap_schedule.fs.schedule_client] keeping b'scheduled-2025-03-12-10_33_00_UTC' due to 6m
2025-03-12T10:33:00.463+0000 7f3b739b2640  0 [snap_schedule DEBUG snap_schedule.fs.schedule_client] keeping b'scheduled-2025-03-12-10_32_01_UTC' due to 6m
2025-03-12T10:33:00.463+0000 7f3b739b2640  0 [snap_schedule DEBUG snap_schedule.fs.schedule_client] keeping b'scheduled-2025-03-12-10_31_00_UTC' due to 6m
2025-03-12T10:33:00.463+0000 7f3b739b2640  0 [snap_schedule DEBUG snap_schedule.fs.schedule_client] keeping b'scheduled-2025-03-12-10_25_04_UTC' due to 6m
2025-03-12T10:33:00.463+0000 7f3b739b2640  0 [snap_schedule DEBUG snap_schedule.fs.schedule_client] keeping b'scheduled-2025-03-12-10_24_00_UTC' due to 6m
```

snap-schedule decided to retain all 5 snapshots since the retention spec is set to 6m which probably didn't expire.

> 
> > The case where mds_server.req_rmsnap_latency.avgcount would still be 0 is when no snapshots are removed. Did this happen in the test? If yes, then that explanation needs to be in the commit message of the fix that change minval to 0 or the other way would be to wait long enough in the test for the snapshot to be removed.
> 
> With respect to check_counters.py - yes. check_counters.py checks if the counters are greater than a particular value (minval). If it's not at the first occurrence, then the test would fail (`not seen`). It doesn't matter what happens next. The `minval` should be fixed, as this issue most probably not hitting due to an earlier [bug](https://tracker.ceph.com/issues/67360). Now that the mismatch is revealed.
> 
> I'll update the commit message with more details and I don't think we need to wait for snapshot to be removed, as check_counters just checks the presence of a counter.

Your change will make the test pass, but I am _more_ interested in the underlying reason for the failure which I mentioned above. The yaml frag for snap-schedule and check-counters is written in a way that ensures that we see snap-schedule removing snapshots according to the retention spec.

So, now, you should check if it suffices to modify the retention spec so that snap-schedule is able to remove snapshots before the test can finish.